### PR TITLE
Fade to black

### PR DESCRIPTION
**Pull request by wsanchez**
This PR sets up the `black` linter.

The current branch does not actually run `black` in the default `tox` config, or in CI, so that the reviewer can see what changes are made *before* we reformat the code base.

To run `black`, use this command: `tox -e black`.  It will fail, since the code is not yet reformatted to comply with `black`'s conventions.

To test out the reformatting, run `tox -e black-reformat`.  At this point, running `black` (per above) should succeed.

Before merging, we need to (1) reformat the code, (2) add `black` to the default `tox` environment list (it's commented out for now) and (3) add the `black` `tox` environment to the CI config (4) remove `W504` from the `flake8` ignore list, as it won't be needed post-reformatting.

### CODE REVIEW

File: MANIFEST.in [hunk 1]
```diff
@@ -1,4 +1,4 @@
-include README.md LICENSE CHANGELOG.md tox.ini requirements-test.txt .coveragerc Makefile pytest.ini .tox-coveragerc
```
Comment (wsanchez) [inline MANIFEST.in]: requirements-test.txt is no longer in the repo

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[build-system]
+
+requires = ["setuptools", "wheel"]
+build-backend = "setuptools.build_meta"
```
Comment (wsanchez) [inline pyproject.toml]: This part is generic config for "use setuptools"

File: tox.ini [hunk 3]
```diff
@@ -1,17 +1,27 @@
 [tox]
 
 envlist =
-    flake8, mypy
+    flake8, mypy  # black
```
Comment (wsanchez) [inline tox.ini]: Change before merge to: flake8, black, mypy

File: tox.ini [hunk 4]
```diff
@@ -31,40 +41,56 @@ basepython =
     pypy3: pypy3
 
 deps =
-    test: coverage==4.5.4
     test: idna==2.8
     test: typing==3.7.4.1
     test: {py26,py27,py34}: pytest==4.6.7
     test: {py35,py36,py37,py38}: pytest==5.2.4
     test: pytest-cov==2.8.1
+    test: {[testenv:coverage_report]deps}
```
Comment (wsanchez) [inline tox.ini]: Use coverage version defined by the `coverage_report` environment. This reduces duplication of this info (and happens to make requires.io work correctly).

File: tox.ini [hunk 5]
```diff
@@ -1,17 +1,27 @@
...
+[default]
+
+basepython = python3.8
+
+setenv =
+    PY_MODULE=hyperlink
+
+    PYTHONPYCACHEPREFIX={envtmpdir}/pycache
+
+
```
Comment (wsanchez) [inline tox.ini]: This is only here for use below as reusable `{[default:XXX]YYY}` values.

File: tox.ini [hunk 6]
```diff
@@ -31,40 +41,56 @@ basepython =
...
-    #   https://help.github.com/en/articles/virtual-environments-for-github-actions#default-environment-variables
-    #   https://docs.travis-ci.com/user/environment-variables#default-environment-variables
-    #   https://www.appveyor.com/docs/environment-variables/
-    codecov: TOXENV CODECOV_* CI
-    codecov: GITHUB_*
-    codecov: TRAVIS TRAVIS_*
-    codecov: APPVEYOR APPVEYOR_*
-
-    # Used in our AppVeyor config
-    codecov: OS
```
Comment (wsanchez) [inline tox.ini]: Moved to `codecov` environment.

File: tox.ini [hunk 7]
```diff
@@ -119,6 +147,9 @@ ignore =
     # variable in global scope should not be mixedCase
     N816,
 
+    # line break before binary operator
+    W503,
+
     # line break after binary operator
     W504,
```
Comment (wsanchez) [inline tox.ini]: Remove before merging

File: tox.ini [hunk 8]
```diff
@@ -188,12 +217,19 @@ ignore_missing_imports = True
 
 description = generate coverage report
 
-basepython = python
+depends = test-py{36,37,38,39,py3}
```
Comment (wsanchez) [inline tox.ini]: This makes parallel `tox` builds work, as it wait for other environments before running `coverage`.

**Comment by wsanchez** [comment]
Restarting builds in hopes that codecov wakes up

**Comment by wsanchez** [comment]
@mahmoud if you get some time for a review, this one would be logistically nice, as this is a difficult branch to merge forward repeatedly due to `black` touching a lot of code, so there are frequent conflicts.

File: tox.ini [hunk 9]
```diff
@@ -61,6 +61,37 @@ commands =
...
+
+
+[testenv:black-reformat]
+
+description  = {[testenv:black]description} and reformat
+basepython   = {[testenv:black]basepython}
+skip_install = {[testenv:black]skip_install}
+deps         = {[testenv:black]deps}
+commands     = {[testenv:black]commands}
+
```
Comment (mahmoud) [inline tox.ini:93]: So this works because the lint args are absent?

File: tox.ini (hunk 8 above)
Comment (mahmoud) [reply tox.ini]: Oh nice, I've been meaning to do this.

File: tox.ini (hunk 9 above)
Comment (wsanchez) [reply tox.ini:93]: Yeah, `{env:BLACK_LINT_ARGS:}` evaluates to `""` in the `black-reformat` environment.

File: tox.ini (hunk 7 above)
Comment (wsanchez) [reply tox.ini]: done

**Comment by wsanchez** [comment]
Thanks for the `W504` reminder; I had let that one slip.

And yes, you should run `tox -e black-reformat` before committing, or at least before pushing. You could also just run `black` directly if you have it installed; some editors will run `black` for you, for example.

I do use the Tox environment, because I know that'll work even if I don't have `black` installed, and it ensured I use the right version in the right way, but YYMV.

**Pull request merged**